source/extensions/http/cache/file_system_http_cache/DESIGN.md

```diff
@@ -0,0 +1,37 @@
+# File system cache design
+
+## Goals
+
+[x] Cache should be usable by two processes at once (e.g. during hot restart)
+[x] Cache should evict the least recently used (LRU) entry when full
+[x] Cache should be limited to a specified amount of storage
+[ ] Cache should mitigate thundering herd problem (i.e. if two or more workers request the same cacheable uncached result at the same time, only one worker should hit upstream). See [discussion](#thundering-herd).
+[ ] There should be an ability to remove objects from the cache with some kind of API call.
+[ ] Cache should expose counters for eviction stats (files evicted, bytes evicted).
+[ ] Cache should expose counters for timing information (eviction thread idle, eviction thread busy)
+[ ] Cache should expose gauges for total size stored.
+[ ] Cache should optionally expose histograms for insert and lookup latencies.
+[x] Cache should index by the request route *and* a key generated from headers that may affect the outcome of a request (See [allowed_vary_headers](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/cache/v3/cache.proto.html))
+[ ] Cache should allow a tree structure of folders, to avoid filesystem performance issues with overcrowded directories.
+
+## Storage design
+
+* The only state stored in memory is that a cache entry is in the process of being written; this allows other requests for the same resource to avoid creating duplicate write operations.
+* The cache can be configured with a maximum number of cache entry files, thereby effectively enforcing a maximum number of files per path.
+* A new cache entry that causes the cache to exceed the configured maximum size or maximum number of entries triggers the eviction thread to evict sufficient LRU entries to bring it back below the threshold\[s\] exceeded.
+* Each cache entry file starts with [a fixed structure header followed by a serialized proto](cache_file_header.proto), followed by proto-serialized headers, raw body and proto-serialized trailers.
+* Cache entry files are named `cache-` followed by a stable hash key for the entry.
+* (When implemented) the tree structure of folders is simply one level deep of folders named `cache-0000`, `cache-0001` etc. as four-digit hexadecimal numbers up to the configured number of subdirectories. Cache files are placed in a folder according to a short stable hash of their key. On cache startup, any cache entries found to be in the wrong folder (as would be the case if the number of folders was reconfigured) will simply be removed.
```

Comment:
It's part of measuring the initial cache size, so does need to be done reasonably early. I have (since this comment) migrated this action to the eviction thread so it doesn't block the main thread that's updating config.